File: poker.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <assert.h>

const int N_CARDS=52;
int CARDS[N_CARDS][2];
const int N_HANDS=5;

int nPatterns=0;
int cntRoyalStraightFlush=0;
int cntStraightFlush=0;
int cntFourOfAKind=0;
int cntFullHouse=0;
int cntFlush=0;
int cntStraight=0;
int cntThreeOfAKind=0;
int cntTwoPair=0;
int cntOnePair=0;
int cntNoPair=0;

bool isRoyalStraightFlush(int hands[N_HANDS][2]){
    if(hands[0][1]!=0){
        return false;
    }
    for(int i=1;i<N_HANDS;i++){
        if(hands[i][0]!=hands[0][0] || hands[i][1]!=i+8){
            return false;
        }
    }
    return true;
}

bool isStraightFlush(int hands[N_HANDS][2]){
    for(int i=1;i<N_HANDS;i++){
        if(hands[i][0]!=hands[0][0] || hands[i][1]!=hands[0][1]+i){
            return false;
        }
    }
    return true;
}

bool isFourOfAKind(int hands[N_HANDS][2]){
    if(hands[0][1]==hands[1][1] && hands[1][1]==hands[2][1] && hands[2][1]==hands[3][1]){
        return true;
    }else if(hands[1][1]==hands[2][1] && hands[2][1]==hands[3][1] && hands[3][1]==hands[4][1]){
        return true;
    }
    return false;
}

bool isFullHouse(int hands[N_HANDS][2]){
    if(hands[0][1]==hands[1][1] && hands[1][1]==hands[2][1] && hands[3][1]==hands[4][1]){
        return true;
    }else if(hands[0][1]==hands[1][1] && hands[2][1]==hands[3][1] && hands[3][1]==hands[4][1]){
        return true;
    }
    return false;
}

bool isFlush(int hands[N_HANDS][2]){
    for(int i=1;i<N_HANDS;i++){
        if(hands[i][0]!=hands[0][0]){
            return false;
        }
    }
    return true;
}

bool isRoyalStraight(int hands[N_HANDS][2]){
    if(hands[0][1]!=0){
        return false;
    }
    for(int i=1;i<N_HANDS;i++){
        if(hands[i][1]!=i+8){
            return false;
        }
    }
    return true;
}

bool isStraight(int hands[N_HANDS][2]){
    if(isRoyalStraight(hands)){
        return true;
    }
    for(int i=1;i<N_HANDS;i++){
        if(hands[i][1]!=hands[0][1]+i){
            return false;
        }
    }
    return true;
}

bool isThreeOfAKind(int hands[N_HANDS][2]){
    if(hands[0][1]==hands[1][1] && hands[1][1]==hands[2][1]){
        return true;
    }else if(hands[1][1]==hands[2][1] && hands[2][1]==hands[3][1]){
        return true;
    }else if(hands[2][1]==hands[3][1] && hands[3][1]==hands[4][1]){
        return true;
    }
    return false;
}

bool isTwoPair(int hands[N_HANDS][2]){
    if(hands[0][1]==hands[1][1] && hands[2][1]==hands[3][1]){
        return true;
    }else if(hands[0][1]==hands[1][1] && hands[3][1]==hands[4][1]){
        return true;
    }else if(hands[1][1]==hands[2][1] && hands[3][1]==hands[4][1]){
        return true;
    }
    return false;
}

bool isOnePair(int hands[N_HANDS][2]){
    for(int i=1;i<N_HANDS;i++){
        if(hands[i-1][1]==hands[i][1]){
            return true;
        }
    }
    return false;
}

void deal(int m, int hands[N_HANDS][2], int k){
    if(m==N_HANDS){
        for(int i=1;i<N_HANDS;i++){
            assert(hands[i-1][1]<=hands[i][1]);
        }
        nPatterns++;
        if(isRoyalStraightFlush(hands)){
            cntRoyalStraightFlush++;
        }else if(isStraightFlush(hands)){
            cntStraightFlush++;
        }else if(isFourOfAKind(hands)){
            cntFourOfAKind++;
        }else if(isFullHouse(hands)){
            cntFullHouse++;
        }else if(isFlush(hands)){
            cntFlush++;
        }else if(isStraight(hands)){
            cntStraight++;
        }else if(isThreeOfAKind(hands)){
            cntThreeOfAKind++;
        }else if(isTwoPair(hands)){
            cntTwoPair++;
        }else if(isOnePair(hands)){
            cntOnePair++;
        }else{
            cntNoPair++;
        }
    }else{
        for(int i=k;i<N_CARDS;i++){
            memcpy(hands[m],CARDS[i],sizeof(CARDS[i]));
            deal(m+1,hands,i+1);
        }
    }
}

int print(const char* str, int cnt){
    return printf("%-21s:%8d(%10.6f%%)\n",str,cnt,100.0*cnt/nPatterns);
}

int main(int argc, char* argv[]){
    for(int i=0;i<N_CARDS;i++){
        CARDS[i][0]=i%4;
        CARDS[i][1]=i/4;
        if(i>0)assert(CARDS[i-1][1]<=CARDS[i][1]);
    }
    int hands[N_HANDS][2];
    deal(0,hands,0);
    print("Royal Straight Flush",cntRoyalStraightFlush);
    print("Straight Flush",cntStraightFlush);
    print("Four of a Kind",cntFourOfAKind);
    print("Full House",cntFullHouse);
    print("Flush",cntFlush);
    print("Straight",cntStraight);
    print("Three of a Kind",cntThreeOfAKind);
    print("Two Pair",cntTwoPair);
    print("One Pair",cntOnePair);
    int len=print("No Pair",cntNoPair);
    for(int i=0;i<len;i++){
        putchar('-');
    }
    putchar('\n');
    print("Total",nPatterns);
    return 0;
}

File: poker.py
JOKER=(0,0)
CARDS=[JOKER]+[(i%4+1,i//4+1) for i in range(52)]

nPatterns=0
cntFiveOfAKind=0
cntRoyalStraightFlush=0
cntStraightFlush=0
cntFourOfAKind=0
cntFullHouse=0
cntFlush=0
cntStraight=0
cntThreeOfAKind=0
cntTwoPair=0
cntOnePair=0
cntNoPair=0

def isFiveOfAKind(hand):
    return hand[0]==JOKER and hand[1][1]==hand[2][1]==hand[3][1]==hand[4][1]

def isRoyalStraightFlush(hand):
    if hand[0]==JOKER:
        if not hand[1][0]==hand[2][0]==hand[3][0]==hand[4][0]:
            return False
        return tuple(hand[i][1] for i in range(1,len(hand))) in [(1,10,11,12),(1,10,11,13),(1,10,12,13),(1,11,12,13),(10,11,12,13)]
    else:
        if hand[0][1]!=1:
            return False
        for i in range(1,len(hand)):
            if hand[i][0]!=hand[0][0] or hand[i][1]!=9+i:
                return False
        return True

def isStraightFlush(hand):
    if hand[0]==JOKER:
        if not hand[1][0]==hand[2][0]==hand[3][0]==hand[4][0]:
            return False
        return tuple(hand[i][1]-hand[1][1] for i in range(2,len(hand))) in [(1,2,3),(1,2,4),(1,3,4),(2,3,4)]
    else:
        for i in range(1,len(hand)):
            if hand[i][0]!=hand[0][0] or hand[i][1]!=hand[0][1]+i:
                return False
        return True

def isFourOfAKind(hand):
    if hand[0]==JOKER:
        return hand[1][1]==hand[2][1]==hand[3][1]\
            or hand[2][1]==hand[3][1]==hand[4][1]
    else:
        return hand[0][1]==hand[1][1]==hand[2][1]==hand[3][1]\
            or hand[1][1]==hand[2][1]==hand[3][1]==hand[4][1]

def isFullHouse(hand):
    if hand[0]==JOKER:
        return hand[1][1]==hand[2][1]==hand[3][1]\
            or hand[2][1]==hand[3][1]==hand[4][1]\
            or hand[1][1]==hand[2][1] and hand[3][1]==hand[4][1]
    else:
        return hand[0][1]==hand[1][1]==hand[2][1] and hand[3][1]==hand[4][1]\
            or hand[0][1]==hand[1][1] and hand[2][1]==hand[3][1]==hand[4][1]

def isFlush(hand):
    j=1 if hand[0]==JOKER else 0
    for i in range(j+1,len(hand)):
        if hand[i][0]!=hand[j][0]:
            return False
    return True

def isRoyalStraight(hand):
    if hand[0]==JOKER:
        return tuple(hand[i][1] for i in range(1,len(hand))) in [(1,10,11,12),(1,10,11,13),(1,10,12,13),(1,11,12,13),(10,11,12,13)]
    else:
        if hand[0][1]!=1:
            return False
        for i in range(1,len(hand)):
            if hand[i][1]!=9+i:
                return False
        return True

def isStraight(hand):
    if isRoyalStraight(hand):
        return True
    if hand[0]==JOKER:
        return tuple(hand[i][1]-hand[1][1] for i in range(2,len(hand))) in [(1,2,3),(1,2,4),(1,3,4),(2,3,4)]
    else:
        for i in range(1,len(hand)):
            if hand[i][1]!=hand[0][1]+i:
                return False
        return True

def isThreeOfAKind(hand):
    if hand[0]==JOKER:
        return hand[1][1]==hand[2][1] or hand[2][1]==hand[3][1] or hand[3][1]==hand[4][1]
    else:
        return hand[0][1]==hand[1][1]==hand[2][1] or hand[1][1]==hand[2][1]==hand[3][1] or hand[2][1]==hand[3][1]==hand[4][1]

def isTwoPair(hand):
    if hand[0]==JOKER:
        return hand[1][1]==hand[2][1] or hand[2][1]==hand[3][1] or hand[3][1]==hand[4][1]
    else:
        return hand[0][1]==hand[1][1] and hand[2][1]==hand[3][1]\
            or hand[0][1]==hand[1][1] and hand[3][1]==hand[4][1]\
            or hand[1][1]==hand[2][1] and hand[3][1]==hand[4][1]

def isOnePair(hand):
    if hand[0]==JOKER:
        return True
    else:
        for i in range(1,len(hand)):
            if hand[i-1][1]==hand[i][1]:
                return True
        return False

def deal(hand, k):
    global nPatterns
    global cntFiveOfAKind
    global cntRoyalStraightFlush
    global cntStraightFlush
    global cntFourOfAKind
    global cntFullHouse
    global cntFlush
    global cntStraight
    global cntThreeOfAKind
    global cntTwoPair
    global cntOnePair
    global cntNoPair

    if len(hand)<5:
        for i in range(k,len(CARDS)):
            hand.append(CARDS[i])
            deal(hand,i+1)
            hand.pop()
    else:
        for i in range(1,len(hand)):
            assert(hand[i-1][1]<=hand[i][1])
        
        nPatterns+=1
        if isFiveOfAKind(hand):
            cntFiveOfAKind+=1
        elif isRoyalStraightFlush(hand):
            cntRoyalStraightFlush+=1
        elif isStraightFlush(hand):
            cntStraightFlush+=1
        elif isFourOfAKind(hand):
            cntFourOfAKind+=1
        elif isFullHouse(hand):
            cntFullHouse+=1
        elif isFlush(hand):
            cntFlush+=1
        elif isStraight(hand):
            cntStraight+=1
        elif isThreeOfAKind(hand):
            cntThreeOfAKind+=1
        elif isTwoPair(hand):
            cntTwoPair+=1
        elif isOnePair(hand):
            cntOnePair+=1
        else:
            cntNoPair+=1 

def output(str, cnt):
    print(f"{str:<21}:{cnt:>8}({100*cnt/nPatterns:10.6f}%)")

if __name__=="__main__":
    deal([],0)
    output("Five of a Kind",cntFiveOfAKind)
    output("Royal Straight Flush",cntRoyalStraightFlush)
    output("Straight Flush",cntStraightFlush)
    output("Four of a Kind",cntFourOfAKind)
    output("Full House",cntFullHouse)
    output("Flush",cntFlush)
    output("Straight",cntStraight)
    output("Three of a Kind",cntThreeOfAKind)
    output("Two Pair",cntTwoPair)
    output("One Pair",cntOnePair)
    output("No Pair",cntNoPair)
    print('-'*len(f"{str():<21}:{int():>8}({float():10.6f}%)"))
    output("Total",nPatterns)